### PR TITLE
Tea in a nutshell part 1: Extension configuration, installation and directory structure

### PR DESCRIPTION
**Pull request by linawolf**
A simple Extbase-based extension. As to not produce an epic pull request only directory structure and extension configuration are explained for now.

Releases: main, 11.5

### CODE REVIEW

File: Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+..  _extbase_tutorial_tea_extension_configuration:
+
+============================
+Create a directory structure
+============================
+
+Extbase requires or defaults to a certain directory structure. It is considered
+best practise to always stick to this structure.
+
+On the first level EXT:tea has the following structure:
```
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst:14]: I  recommend to link EXT:tea to the GitHub repository.

File: Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+..  code-block:: none
+    :caption: Directory structure of EXT:tea
+
+    $ tree typo3conf/ext/tea
```
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst:19]: We should avoid `typo3conf/ext` as this is gone for Composer-based extensions using typo3/cms-composer-installers v4 (available as RC for TYPO3 v11 and AFAIK v12 relies on it). The extension is then installed under `vendors/ttn/tea` with a symlink or `Resources/Public` to `_assets/<some hash>/`

File: Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    ├── Tests
+    ├── composer.json
+    ├── ext_emconf.php
+    ├── ...
+    └── README.md
+
+Directory :file:`Classes`
+-------------------------
+
+The folder :file:`Classes` has to contain all PHP classes provided by the
```
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst:33]: Hmm. Maybe "should contain" instead of "has to contain" and further on "not available in the [default|recommended] autoloading". In can also use "src" folder, but have to adjust the autoloading ;-)
Or "has to contain all PHP classes by convention".
Just as a food for thought ;-)

File: Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+extension. Otherwise they are not available in the
+:ref:`autoloading <autoload>`.
+
+See also the general chapter on the folder :ref:`extension-classes`.
+
+In the :file:`composer.json` we defined that all PHP classes should
```
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst:39]:
```suggestion
In the :file:`composer.json` we have defined that all PHP classes are
```

File: Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+automatically be loaded from the directory :file:`Classes`:
```
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst:40]:
```suggestion
automatically loaded from the :file:`Classes` directory:
```

File: Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+
+            {
+                "autoload": {
+                    "psr-4": {
+                        "TTN\\Tea\\": "Classes/"
+                    }
+                }
+            }
+
+    ..  group-tab:: Legacy
```
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst:58]: This example does not match with the sentence above ("In the :file:`composer.json` we defined") - why then ext_emconf.php?

File: Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+
+            $EM_CONF[$_EXTKEY] = [
+                'autoload' => [
+                    'psr-4' => [
+                        'TTN\\Tea\\' => 'Classes/',
+                    ],
+                ],
+            ];
+
+Further this line defines the namespace all PHP classes in this directory
```
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst:72]: This sentence sounds a bit bumpy.

File: Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+have to be situated in to be found by the :ref:`PSR-4 autoloading <autoload>`.
+
+The folder :file:`Classes` contains several subfolders:
+
+..  code-block:: none
+    :caption: Directory structure of EXT:tea
+
+    $ tree typo3conf/ext/tea
```
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst:80]: Here also: avoid `typo3conf/ext`.

File: Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+-------------------------------
+
+See also the general chapter on the folder :ref:`extension-configuration-files`.
+
+The folder :file:`Configuration` contains several subfolders:
+
+..  code-block:: none
+    :caption: Directory structure of EXT:tea
+
+    $ tree typo3conf/ext/tea
```
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst:112]: Here also: avoid `typo3conf/ext`.

File: Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+        |   ├── Page
+        |   └── User
+        ├── TypoScript
+        |   ├── constants.typoscript
+        |   └── setup.typoscript
+        └──  Services.yaml
+
+:file:`Configuration/FlexForms`
+    Contains the configuration of additional input fields to
+    configure the plugin.
```
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst:127]:
```suggestion
    configure plugins.
```
Without switchableControllerActions you might have several plugins now. Would also like to see a link to FlexForm explanation.

File: Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+:file:`Configuration/TCA`
+    Contains the TYPO3 configuration array (TCA) as PHP arrays.
```
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst:129]: Would like to see a link to TCA as quick reference.

File: Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+:file:`Configuration/TCA/Overrides`
+    Can be used to extend the TCA of other extensions. They can be extended
+    by direct array manipulation or (preferred if possible) by calls to
+    API functions.
+:file:`Configuration/TsConfig`
+    Contains configurations for the TYPO3 backend on page or user level in
```
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst:135]: Would like to see links to page TSconfig and user TSconfig as quick reference.

File: Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+    the syntax TypoScript. This extension does not feature TS
+:file:`Configuration/TypoScript`
+    Contains configurations for the TYPO3 frontend. In some contexts the
```
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst:138]: Would like to see a link to TypoScript reference.

File: Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+    configuration contained here is also considered in the backend.
+:file:`Configuration/Services.yaml`
+    Is used to configure technical aspects of the extension including
```
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst:141]: Would like to see a link to DI chapter as quick reference.

File: Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+
+See also the general chapter on the folder :ref:`extension-Resources`.
+
+The folder :file:`Resources` contains two sub folders that are
+further divided:
+
+..  code-block:: none
+    :caption: Directory structure of EXT:tea
+
+    $ tree typo3conf/ext/tea
```
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst:162]: Here also: avoid `typo3conf/ext`.

File: Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+        |   └── Templates
+        └── Public
+            ├── CSS
+            ├── Icons
+            ├── Images
+            └── JavaScript
+
+
+:file:`Resources/Private`
+    All ressource files that do not have to be loaded directly by a browser
```
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst:177]:
```suggestion
    All resource files that do not have to be loaded directly by a browser
```

Maybe: "... must not be accessible by a browser ..."
This distinction is more necessary with typo3/cms-composer-installers v4.

File: Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+    **should** go in this directory. This includes Fluid templating files
```
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst:178]:
```suggestion
    **must** go in this directory. This includes Fluid templating files.
```

File: Documentation/ExtensionArchitecture/Tutorials/Tea/ExtensionConfiguration.rst [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+        ..  include:: /CodeSnippets/Tutorials/Tea/ComposerJsonSimplified.rst.txt
+
+    ..  group-tab:: Legacy
+
+        ..  include:: /CodeSnippets/Tutorials/Tea/ExtEmconf.rst.txt
+
+
+..  hint::
+    If the extension should also be available for legacy installations it
+    also needs a file called :file:`EXT:tea/ext_emconf.php`. This file
```
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/ExtensionConfiguration.rst:31]:
```suggestion
    also needs a file called :file:`ext_emconf.php`. This file
```
Would avoid EXT:tea as it is also not mentioned above when mentioning composer.json.

File: Documentation/ExtensionArchitecture/Tutorials/Tea/ExtensionConfiguration.rst [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+    contains roughly the same information in a different format. Have a look
+    at the tab "Legacy" above.
+
+With just this file present you would be able to install the extension
```
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/ExtensionConfiguration.rst:35]: Would rephrase it: "With just the composer.json present (and for legacy installations additionally ext_emconf.php) ..."

File: Documentation/ExtensionArchitecture/Tutorials/Tea/ExtensionConfiguration.rst [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+but it would not do anything.
+
+Though not required it is considered best practice for an extension to have an
+icon. This icon should have the format :file:`.svg` or :file:`.png` and has
+to be located at :file:`EXT:tea/Resources/Public/Icons/Extension.svg`.
```
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/ExtensionConfiguration.rst:40]:
```suggestion
to be located at :file:`EXT:tea/Resources/Public/Icons/Extension.svg` by convention.
```

File: Documentation/ExtensionArchitecture/Tutorials/Tea/Index.rst [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+
+    Create the files needed to have a minimal running extension and install it.
+
+#.  :ref:`Directory structure <extbase_tutorial_tea_extension_configuration>`
+
+    Have a look at the directory structure of the example extension and learn
+    which files should go where.
+
+..  note::
+    This tutorial is work in progress. As we want to prevent epic pull requests,
```
Comment (brotkrueml) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/Index.rst:43]: I like this :-)

**Review by brotkrueml — CHANGES_REQUESTED** [review]
Just some random comments ;-) I haven't rendered it locally by now, just reviewed here. Will do it later after my comments are solved. I recommend that Tom has a look of the wording after it matured. Some sentences are bumpy.

File: Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst (hunk 2 above)
Comment (linawolf) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst:19]:
```suggestion
    $ tree /path/to/extension/tea
```

File: Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst (hunk 3 above)
Comment (linawolf) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst:33]:
```suggestion
The folder :file:`Classes` should contain all PHP classes provided by the
```

File: Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst [hunk 22]
```diff
@@ -0,0 +1,189 @@
...
+    ├── composer.json
+    ├── ext_emconf.php
+    ├── ...
+    └── README.md
+
+Directory :file:`Classes`
+-------------------------
+
+The folder :file:`Classes` has to contain all PHP classes provided by the
+extension. Otherwise they are not available in the
```
Comment (linawolf) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst:34]:
```suggestion
extension. Otherwise they are not available in the default
```

File: Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst [hunk 23]
```diff
@@ -0,0 +1,189 @@
...
+
+Further this line defines the namespace all PHP classes in this directory
+have to be situated in to be found by the :ref:`PSR-4 autoloading <autoload>`.
+
+The folder :file:`Classes` contains several subfolders:
+
+..  code-block:: none
+    :caption: Directory structure of EXT:tea
+
+    $ tree typo3conf/ext/tea
```
Comment (linawolf) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst:80]:
```suggestion
    $ tree path/to/extension/tea
```

File: Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst [hunk 24]
```diff
@@ -0,0 +1,189 @@
...
+-------------------------------
+
+See also the general chapter on the folder :ref:`extension-configuration-files`.
+
+The folder :file:`Configuration` contains several subfolders:
+
+..  code-block:: none
+    :caption: Directory structure of EXT:tea
+
+    $ tree typo3conf/ext/tea
```
Comment (linawolf) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst:112]:
```suggestion
    $ tree path/to/extension/tea
```

File: Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst [hunk 25]
```diff
@@ -0,0 +1,189 @@
...
+
+See also the general chapter on the folder :ref:`extension-Resources`.
+
+The folder :file:`Resources` contains two sub folders that are
+further divided:
+
+..  code-block:: none
+    :caption: Directory structure of EXT:tea
+
+    $ tree typo3conf/ext/tea
```
Comment (linawolf) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst:162]:
```suggestion
    $ tree /path/to/extension/tea
```

File: Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst (hunk 17 above)
Comment (linawolf) [reply Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst:178]: Is it? Some people put for example  scss files in the public folder even though they don't have to be accesssible by the browser? Main rule is: When we use **must** we describe what happens otherwise

File: Documentation/ExtensionArchitecture/Tutorials/Tea/ExtensionConfiguration.rst (hunk 19 above)
Comment (linawolf) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/ExtensionConfiguration.rst:35]:
```suggestion
With just the :file:`composer.json` present (and for legacy installations additionally 
:file:`ext_emconf.php`) you would be able to install the extension
```

File: Documentation/ExtensionArchitecture/Tutorials/Tea/ExtensionConfiguration.rst (hunk 20 above)
Comment (linawolf) [reply Documentation/ExtensionArchitecture/Tutorials/Tea/ExtensionConfiguration.rst:40]: Is this a convention? From What I know it is not found unless it is placed here?

File: Documentation/ExtensionArchitecture/Tutorials/Tea/Index.rst (hunk 21 above)
Comment (linawolf) [reply Documentation/ExtensionArchitecture/Tutorials/Tea/Index.rst:43]: :)

File: Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst (hunk 6 above)
Comment (linawolf) [reply Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst:59]: Fixed

File: Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst (hunk 14 above)
Comment (linawolf) [reply Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst:146]: fixed

File: Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst (hunk 17 above)
Comment (brotkrueml) [reply Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst:184]: With typo3/cms-composer-installers v4 everything in Resources/Public is accessible by browser. As it is also with v3 as it is not restricted by .htaccess. I would say so, yes. If you want to be more polite say "strongly recommended". I want to emphasize that only accessible artifacts should be placed in Public.
Remark: I also heard from people who placed PHP scripts into Public :-(.

File: Documentation/ExtensionArchitecture/Tutorials/Tea/ExtensionConfiguration.rst (hunk 20 above)
Comment (brotkrueml) [reply Documentation/ExtensionArchitecture/Tutorials/Tea/ExtensionConfiguration.rst:41]: Hmm, yes, that's the convention. But you are right, maybe "has to be located" is strong enough. Better "must be located" ;-) But that is nitpicking.

File: Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst (hunk 17 above)
Comment (linawolf) [reply Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst:184]: There are also installation that have an Assets folder and then some Building processes that put all publicly available files into public. that is also legit...

File: Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst [hunk 26]
```diff
@@ -0,0 +1,199 @@
...
+..  _extbase_tutorial_tea_extension_configuration:
+
+============================
+Create a directory structure
+============================
+
+Extbase requires or defaults to a certain directory structure. It is considered
+best practise to always stick to this structure.
+
+On the first level EXT:tea has the following structure:
```
Comment (linawolf) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/DirectoryStructure.rst:14]:
```suggestion
On the first level `EXT:tea <https://github.com/TYPO3-Documentation/tea>`__ has the following structure:
```

File: Documentation/ExtensionArchitecture/Tutorials/Tea/Index.rst [hunk 27]
```diff
@@ -0,0 +1,50 @@
...
+The example extension :t3ext:`tea` was created as an example of best practises
+on automatic code checks.
+
+..  hint::
+    If you want to learn more about automatic code checks
+    see the :doc:`documentation of tea <ext_tea:Index>` and the chapter on
+    :ref:`Testing <testing>` in this manual.
+
+In this manual, however we will ignore the testing and just explain how this
+simple and very well tested example extension works.
```
Comment (linawolf) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/Index.rst:22]:
```suggestion
 example extension works. The extension demonstrates basic functionality and is very well tested
```

File: Documentation/ExtensionArchitecture/Tutorials/Tea/Index.rst [hunk 28]
```diff
@@ -0,0 +1,50 @@
...
+    If you want to learn more about automatic code checks
+    see the :doc:`documentation of tea <ext_tea:Index>` and the chapter on
+    :ref:`Testing <testing>` in this manual.
+
+In this manual, however we will ignore the testing and just explain how this
+ example extension works. The extension demonstrates basic functionality and is very well tested
+
+..  attention::
+    The example extension tea should not be used as a kickstarter or
+    template for your own extension. It is simply an example to be
```
Comment (linawolf) [inline Documentation/ExtensionArchitecture/Tutorials/Tea/Index.rst:26]:
```suggestion
    template for your own extension. It is an example to be
```

**Review by DavidBruchmann — APPROVED** [review]
